pkg/tools/ha/lease.go

```diff
@@ -0,0 +1,268 @@
+// It also is used to track the state of a program (i.e. if lease obtained, Running should be true)
+type LeaseUser struct {
+
+	//Note: In general 'Running' should be 'false' when you first start an HA process, and it should
+	//be flipped to 'true' once you acquire the lock and start up.
+	Running      bool // This is how users verify that their process is running.
+	LeasesGained int  // Number of times lease has been granted.
+	LeasesLost   int  // Number of times the lease was lost.
+}
+
+// Configuration options needed for running the lease loop.
+type Config struct {
+	Key    string
+	whoami string
+	ttl    uint64
+	sleep  time.Duration
+	// These two functions return "err" or else nil.
+	// They should also update information in the LeaseUserInfo struct
+	// about the state of the lease owner.
+	LeaseGained   func(lu *LeaseUser) bool
+	LeaseLost     func(lu *LeaseUser) bool
+	Cli           *client.Client
+	LeaseUserInfo *LeaseUser
+}
+
+func (c *Config) AddFlags(fs *pflag.FlagSet) {
+	//If key or whoami aren't provided, they should be set by default by the calling program.
+	pflag.StringVar(&c.Key, "key", "", "The key to use for the lock")
+	pflag.StringVar(&c.whoami, "whoami", "", "The name to use for the reservation.  If empty os.Hostname is the default.")
+	pflag.Uint64Var(&c.ttl, "ttl", 30, "The time to live for the lock.")
+	pflag.DurationVar(&c.sleep, "sleep", 5*time.Second, "The length of time to sleep between checking the lock.")
+}
+
+// RunHA runs a process in a highly available fashion.
+func RunHA(kubecfg string, master string, start func(l *LeaseUser) bool, stop func(l *LeaseUser) bool, mcfg *Config) {
+
+	//We need a kubeconfig in order to use the locking API, so we create it here.
+	kubeconfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubecfg},
+		&clientcmd.ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: master}}).ClientConfig()
+
+	if err != nil {
+		glog.Infof("Exiting, couldn't create kube configuration with parameters cfg=%v and master=%v ", kubecfg, master)
+		os.Exit(1)
+	}
+
+	kubeClient, err := client.New(kubeconfig)
+
+	leaseUserInfo := LeaseUser{
+		Running:      false,
+		LeasesGained: 0,
+		LeasesLost:   0,
+	}
+
+	mcfg.LeaseUserInfo = &leaseUserInfo
+	mcfg.LeaseGained = start
+	mcfg.LeaseLost = stop
+	mcfg.Cli = kubeClient
+
+	RunLease(mcfg)
+	glog.Infof("zz Done with starting the lease loop for %v.", mcfg)
+}
+
+var countsleeps = 0
+
+// leaseAndUpdateLoop runs the loop to acquire a lease.  This will continue trying to get a lease
+// until it succeeds, and then callbacks sent in by the user will be triggered.
+// Likewise, it can give up the lease, in which case the LeaseLost() callbacks will be triggered.
+func (c *Config) leaseAndUpdateLoop() {
+	for {
+		//Proactively give up the lease every 50 tries.
+		var master bool = false
+		var err error = nil
+
+		//TODO Delete this condition, its only to cause quick failover.
```

Comment:
This needs to be removed before we merge.  I've left it in b/c i think a timebomb might be useful, so I can evolve it if you folks want to keep it.  otherwise ill just delete entirely.